File: internal/bot/command/add_google_app.go
package command

import (
	"fmt"

	tgbotapi "github.com/go-telegram-bot-api/telegram-bot-api/v5"
	"github.com/miti99/store-scraper-bot-go/internal/api/google"
	"github.com/miti99/store-scraper-bot-go/internal/api/google/request"
	"github.com/miti99/store-scraper-bot-go/internal/config"
	"github.com/miti99/store-scraper-bot-go/internal/repository"
)

// /addgoogle <appId> [country=vn] — Java AddGoogleAppCommand.
type AddGoogleAppCommand struct {
	cfg       *config.Config
	adminRepo *repository.AdminRepository
	groupRepo *repository.GroupRepository
	scraper   *google.GoogleScraper
}

func NewAddGoogleAppCommand(cfg *config.Config, adminRepo *repository.AdminRepository, groupRepo *repository.GroupRepository, scraper *google.GoogleScraper) *AddGoogleAppCommand {
	return &AddGoogleAppCommand{cfg: cfg, adminRepo: adminRepo, groupRepo: groupRepo, scraper: scraper}
}

func (c *AddGoogleAppCommand) Execute(msg *tgbotapi.Message, sender Sender) {
	if !authorizeGroup(msg.Chat.ID, c.adminRepo, sender) {
		return
	}
	args := splitArgs(msg.CommandArguments())
	if len(args) < 1 || len(args) > 2 {
		_ = sender.SendMessage(msg.Chat.ID, "Invalid arguments")
		return
	}
	appID := args[0]
	country := "vn"
	if len(args) == 2 {
		country = args[1]
	}
	resp, err := c.scraper.FetchAndCache(request.New(appID, country))
	if err != nil || resp == nil {
		_ = sender.SendMessage(msg.Chat.ID, "Error when request app info")
		return
	}
	added, err := c.groupRepo.AddGoogleApp(msg.Chat.ID, appID, country)
	if err != nil {
		_ = sender.SendMessage(msg.Chat.ID, "Internal server error")
		return
	}
	if !added {
		_ = sender.SendMessage(msg.Chat.ID, fmt.Sprintf("Google app <code>%s</code> is already added", appID))
		return
	}
	_ = sender.SendMessage(msg.Chat.ID, fmt.Sprintf("Google app <code>%s</code>, country <b>%s</b> added successfully", appID, country))
}

File: src/bot/commands/add-google-app.js
import { buildGoogleRequest } from '../../api/google-scraper.js';
import * as groupRepo from '../../repository/group-repository.js';
import { authorizeGroup, getCommandArguments, splitArgs } from './command-utils.js';

// /addgoogle <appId> [country=vn] — Java AddGoogleAppCommand.
export function createAddGoogleAppCommand(googleScraper) {
  return async (msg, sender) => {
    if (!(await authorizeGroup(msg.chat.id, sender))) return;
    const args = splitArgs(getCommandArguments(msg.text));
    if (args.length < 1 || args.length > 2) {
      await sender.sendMessage(msg.chat.id, 'Invalid arguments');
      return;
    }
    const appId = args[0];
    const country = args.length === 2 ? args[1] : 'vn';

    let resp;
    try {
      resp = await googleScraper.fetchAndCache(buildGoogleRequest(appId, country));
    } catch {
      resp = null;
    }
    if (!resp) {
      await sender.sendMessage(msg.chat.id, 'Error when request app info');
      return;
    }

    try {
      const added = await groupRepo.addGoogleApp(msg.chat.id, appId, country);
      if (!added) {
        await sender.sendMessage(msg.chat.id, `Google app <code>${appId}</code> is already added`);
        return;
      }
      await sender.sendMessage(
        msg.chat.id,
        `Google app <code>${appId}</code>, country <b>${country}</b> added successfully`,
      );
    } catch {
      await sender.sendMessage(msg.chat.id, 'Internal server error');
    }
  };
}
